File: services/node/src/cache.ts

import Redis from "ioredis";
import type { Config } from "./config";

export function createRedis(cfg: Config): Redis {
  return new Redis(cfg.redisUrl, {
    lazyConnect: true,
    maxRetriesPerRequest: 2,
    enableReadyCheck: true,
  });
}

const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms));

/**
 */
export async function connectWithRetry(
  redis: Redis,
  log: (msg: string) => void,
  maxMs = 30_000,
): Promise<void> {
  const deadline = Date.now() + maxMs;
  let attempt = 0;
  while (true) {
    attempt += 1;
    try {
      if (redis.status !== "ready") {
        if (redis.status === "wait" || redis.status === "end") {
          await redis.connect();
        }
      }
      await redis.ping();
      log(`cache: connected (attempt ${attempt})`);
      return;
    } catch (err) {
      if (Date.now() >= deadline) {
        throw new Error(
          `cache: could not connect within ${maxMs}ms: ${(err as Error).message}`,
        );
      }
      log(
        `cache: not ready (attempt ${attempt}), retrying: ${(err as Error).message}`,
      );
      await sleep(1000);
    }
  }
}

export async function cacheHealthy(redis: Redis): Promise<boolean> {
  try {
    const pong = await redis.ping();
    return pong === "PONG";
  } catch {
    return false;
  }
}

export function productKey(stack: string, id: string | number): string {
  return `${stack}:product:${id}`;
}

File: services/rust/src/cache.rs
use std::time::Duration;

use redis::aio::MultiplexedConnection;
use redis::AsyncCommands;
use tokio::time::sleep;

#[derive(Clone)]
pub struct Cache {
    conn: MultiplexedConnection,
    prefix: String,
}

impl Cache {
    pub async fn connect(url: &str, stack: &str) -> Result<Cache, redis::RedisError> {
        let client = redis::Client::open(url)?;
        let deadline = std::time::Instant::now() + Duration::from_secs(30);
        let mut last_err: Option<redis::RedisError> = None;

        loop {
            match client.get_multiplexed_async_connection().await {
                Ok(conn) => {
                    tracing::info!("connected to redis");
                    return Ok(Cache { conn, prefix: format!("{stack}:") });
                }
                Err(e) => {
                    if std::time::Instant::now() >= deadline {
                        return Err(last_err.unwrap_or(e));
                    }
                    tracing::warn!("redis not ready, retrying in 1s: {e}");
                    last_err = Some(e);
                    sleep(Duration::from_secs(1)).await;
                }
            }
        }
    }

    fn key(&self, suffix: &str) -> String {
        format!("{}{}", self.prefix, suffix)
    }

    pub async fn get(&self, suffix: &str) -> Result<Option<String>, redis::RedisError> {
        let mut conn = self.conn.clone();
        conn.get(self.key(suffix)).await
    }

    pub async fn set_ex(
        &self,
        suffix: &str,
        value: &str,
        ttl_seconds: u64,
    ) -> Result<(), redis::RedisError> {
        let mut conn = self.conn.clone();
        conn.set_ex(self.key(suffix), value, ttl_seconds).await
    }

    pub async fn del(&self, suffix: &str) -> Result<(), redis::RedisError> {
        let mut conn = self.conn.clone();
        conn.del::<_, ()>(self.key(suffix)).await
    }

    pub async fn ping(&self) -> Result<(), redis::RedisError> {
        let mut conn = self.conn.clone();
        redis::cmd("PING").query_async::<()>(&mut conn).await
    }
}
